flink-clients/src/main/java/org/apache/flink/client/python/PythonDriver.java

```diff
@@ -0,0 +1,170 @@
+		} finally {
+			gatewayServer.shutdown();
+		}
+	}
+
+	/**
+	 * Creates a GatewayServer run in a daemon thread.
+	 *
+	 * @return The created GatewayServer
+	 */
+	public static GatewayServer startGatewayServer() {
+		InetAddress localhost = InetAddress.getLoopbackAddress();
+		GatewayServer gatewayServer = new GatewayServer.GatewayServerBuilder()
+			.javaPort(0)
+			.javaAddress(localhost)
+			.build();
+		Thread thread = new Thread(gatewayServer::start);
+		thread.setName("py4j-gateway-init");
+		thread.setDaemon(true);
+		thread.start();
+		try {
+			thread.join();
+		} catch (InterruptedException e) {
+			LOG.error("The gateway server thread join failed.", e);
+			System.exit(1);
+		}
+		return gatewayServer;
+	}
+
+	/**
+	 * Constructs the commands which will be executed in python process.
+	 *
+	 * @param filePathMap stores python file name to its path
+	 * @param parsedArgs  parsed args
+	 */
+	public static List<String> constructCommands(Map<String, Path> filePathMap, Map<String, List<String>> parsedArgs) {
+		List<String> commands = new ArrayList<>();
+		if (parsedArgs.containsKey("python")) {
+			String pythonFile = parsedArgs.get("python").get(0);
+			Path pythonFilePath = new Path(pythonFile);
+			filePathMap.put(pythonFilePath.getName(), pythonFilePath);
+			commands.add(pythonFilePath.getName());
+		}
+		if (parsedArgs.containsKey("pyModule")) {
+			String pyModule = parsedArgs.get("pyModule").get(0);
+			commands.add("-m");
+			commands.add(pyModule);
+		}
+		if (parsedArgs.containsKey("pyFiles")) {
+			List<String> pyFiles = parsedArgs.get("pyFiles");
+			for (String pyFile : pyFiles) {
+				Path pyFilePath = new Path(pyFile);
+				filePathMap.put(pyFilePath.getName(), pyFilePath);
+			}
+		}
+		if (parsedArgs.containsKey("args")) {
+			commands.addAll(parsedArgs.get("args"));
+		}
+		return commands;
+	}
+
+	/**
+	 * Parses the args to the map format.
+	 *
+	 * @param args ["python", "xxx.py",
+	 *             "pyFiles", "a.py,b.py,c.py",
+	 *             "--input", "in.txt"]
+	 * @return {"python"->List("xxx.py"),"pyFiles"->List("a.py","b.py","c.py"),"args"->List("--input","in.txt")}
+	 */
+	public static Map<String, List<String>> parseOptions(String[] args) {
+		Map<String, List<String>> parsedArgs = new HashMap<>();
+		int argIndex = 0;
+		boolean isValidPythonFile = false;
+		// valid args should include python or pyModule field and their value.
+		if (args[0].equals("python") || args[0].equals("pyModule")) {
+			parsedArgs.put(args[0], Collections.singletonList(args[1]));
+			argIndex = 2;
+			isValidPythonFile = true;
+		}
+		if (isValidPythonFile && args.length > 2 && args[2].equals("pyFiles")) {
+			List<String> pyFilesList = new ArrayList<>(Arrays.asList(args[3].split(",")));
+			parsedArgs.put(args[2], pyFilesList);
+			argIndex = 4;
+		}
+		if (!isValidPythonFile) {
+			throw new RuntimeException("Args is invalid, the args is required to include python main file or pyModule");
```

Comment:
exception msg: The Python entrypoint has not been specified. It can be specified with option -py or -pym.